qiskit/algorithms/minimum_eigen_solvers/adapt_vqe.py

```diff
@@ -0,0 +1,336 @@
+
+    `AdaptVQE <https://arxiv.org/abs/1812.11173>`__ is a quantum algorithm creates a compact ansatz
+    by gradually building up the ansatz circuit by appending the excitation with the largest energy
+    gradient to the circuit. This results in a wavefunction ansatz that is uniquely formed by the
+    algorithm.
+    This is an adaptive wrapper of a VQE used internally as solver.
+    The performance of AdaptVQE can significantly depend on the choice of gradient method, QFI
+    solver (if applicable) and the epsilon value.
+    """
+
+    def __init__(
+        self,
+        solver: VQE,
+        adapt_gradient: Optional[GradientBase] = None,
+        excitation_pool: List[Union[OperatorBase, QuantumCircuit]] = None,
+        threshold: float = 1e-5,
+        max_iterations: Optional[int] = None,
+    ) -> None:
+        """
+        Args:
+            solver: a factory for the VQE solver employing a UCCSD ansatz.
+            adapt_gradient: a class that converts operator expression to the first-order gradient based
+                on the method mentioned.
+            excitation_pool: An entire list of excitations.
+            threshold: the energy convergence threshold. It has a minimum value of 1e-15.
+            max_iterations: the maximum number of iterations of the AdaptVQE algorithm.
+        """
+        validate_min("threshold", threshold, 1e-15)
+
+        if adapt_gradient is None:
+            adapt_gradient = Gradient(grad_method="param_shift")
+        self._threshold = threshold
+        self.solver = solver
+        self._tmp_ansatz = self.solver.ansatz
+        self._max_iterations = max_iterations
+        self._adapt_gradient = adapt_gradient
+        self._excitation_pool = excitation_pool
+        self._excitation_list: List[OperatorBase] = []
+
+    @property
+    def initial_point(self) -> Optional[np.ndarray]:
+        """Returns initial point."""
+        return self.solver.initial_point
+
+    @initial_point.setter
+    def initial_point(self, initial_point: Optional[np.ndarray]) -> None:
+        """Sets initial point."""
+        self.solver.initial_point = initial_point
+
+    def _compute_gradients(
+        self,
+        theta: List[float],
+        operator: OperatorBase,
+    ) -> List[Tuple[float, PauliSumOp]]:
+        """
+        Computes the gradients for all available excitation operators.
+
+        Args:
+            theta: List of (up to now) optimal parameters
+            operator: operator whose gradient needs to be computed
+        Returns:
+            List of pairs consisting of gradient and excitation operator.
+        Raises:
+            AlgorithmError: If `quantum_instance` is not provided.
+        """
+        res = []
+        # compute gradients for all excitation in operator pool
+        if self.solver.quantum_instance is None:
+            raise AlgorithmError(
+                "A QuantumInstance or Backend must be supplied to run the quantum algorithm."
+            )
```

Comment:
I think we should extract this check to be done only once at the beginning of `compute_minimum_eigenvalue` instead of at every iteration. The docstrings should be updated accordingly.